internal/controller/dns/record.go

```diff
@@ -0,0 +1,223 @@
+		WithOptions(o).
+		For(&v1alpha1.Record{}).
+		Complete(r)
+}
+
+// A connector is expected to produce an ExternalClient when its Connect method
+// is called.
+type connector struct {
+	kube                  client.Client
+	newCloudflareClientFn func(cfg clients.Config) records.Client
+}
+
+// Connect produces a valid configuration for a Cloudflare API
+// instance, and returns it as an external client.
+func (c *connector) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	_, ok := mg.(*v1alpha1.Record)
+	if !ok {
+		return nil, errors.New(errNotDNSRecord)
+	}
+
+	// Get client configuration
+	config, err := clients.GetConfig(ctx, c.kube, mg)
+	if err != nil {
+		return nil, errors.Wrap(err, errClientConfig)
+	}
+
+	return &external{client: c.newCloudflareClientFn(*config)}, nil
+}
+
+// An ExternalClient observes, then either creates, updates, or deletes an
+// external resource to ensure it reflects the managed resource's desired state.
+type external struct {
+	client records.Client
+}
+
+func (e *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mg.(*v1alpha1.Record)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotDNSRecord)
+	}
+
+	en := meta.GetExternalName(cr)
+
+	// If the name & ExternalName are the same we know we haven't set the ExternalName
+	// So attempt to create
+	if en == cr.ObjectMeta.Name {
+		return managed.ExternalObservation{ResourceExists: false}, nil
+	}
+
+	if cr.Spec.ForProvider.Zone == nil {
+		return managed.ExternalObservation{}, errors.New(errDNSRecordNoZone)
+	}
+
+	record, err := e.client.DNSRecord(ctx, *cr.Spec.ForProvider.Zone, en)
+
+	if err != nil {
+		// Been deleted or doesnt exist
+		if records.IsRecordNotFound(err) {
+			return managed.ExternalObservation{ResourceExists: false}, nil
+		}
+		return managed.ExternalObservation{}, errors.Wrap(err, errDNSRecordLookup)
+	}
+
+	cr.Status.AtProvider = records.GenerateObservation(record)
+
+	return managed.ExternalObservation{
+		ResourceExists:          true,
+		ResourceUpToDate:        records.UpToDate(&cr.Spec.ForProvider, record),
+		ResourceLateInitialized: records.LateInitialize(&cr.Spec.ForProvider, record),
+	}, nil
+}
+
+func (e *external) Create(ctx context.Context, mg resource.Managed) (managed.ExternalCreation, error) {
+	cr, ok := mg.(*v1alpha1.Record)
+	if !ok {
+		return managed.ExternalCreation{}, errors.New(errNotDNSRecord)
+	}
+
+	if cr.Spec.ForProvider.Zone == nil {
+		return managed.ExternalCreation{}, errors.New(errDNSRecordCreation)
+	}
+
+	if cr.Spec.ForProvider.TTL == nil {
+		return managed.ExternalCreation{}, errors.New(errDNSRecordCreation)
+	}
+
+	// TODO: Add validation here for priority (only required for specific record types)
+
+	cr.SetConditions(rtv1.Creating())
+
+	res, err := e.client.CreateDNSRecord(
+		ctx,
+		*cr.Spec.ForProvider.Zone,
+		cloudflare.DNSRecord{
+			Type:     *cr.Spec.ForProvider.Type,
```

Comment:
`Type` is defaulted in the API spec but probably still smart to validate it like the `Zone` and `TTL` field above, just in case. Better to have an additional nil pointer check than panic because `Type` wasn't set.